pace-util/pace/util/decomposition.py

```diff
@@ -0,0 +1,141 @@
+
+
+def compiling_equivalent(rank: int, partitioner: TilePartitioner):
+    """From my rank & the current partitioner we determine which
+    rank we should read from"""
+    if partitioner.layout == (1, 1):
+        return 0
+    if partitioner.layout == (2, 2):
+        if partitioner.tile.on_tile_bottom(rank):
+            if partitioner.tile.on_tile_left(rank):
+                return 0  # "00"
+            if partitioner.tile.on_tile_right(rank):
+                return 1  # "10"
+        if partitioner.tile.on_tile_top(rank):
+            if partitioner.tile.on_tile_left(rank):
+                return 2  # "01"
+            if partitioner.tile.on_tile_right(rank):
+                return 3  # "11"
+    if partitioner.layout == (3, 3):
+        if partitioner.tile.on_tile_bottom(rank):
+            if partitioner.tile.on_tile_left(rank):
+                return 0  # "00"
+            if partitioner.tile.on_tile_right(rank):
+                return 2  # "20"
+            else:
+                return 1  # "10"
+        if partitioner.tile.on_tile_top(rank):
+            if partitioner.tile.on_tile_left(rank):
+                return 6  # "02"
+            if partitioner.tile.on_tile_right(rank):
+                return 8  # "22"
+            else:
+                return 7  # "12"
+        else:
+            if partitioner.tile.on_tile_left(rank):
+                return 3  # "01"
+            if partitioner.tile.on_tile_right(rank):
+                return 5  # "21"
+            else:
+                return 4  # "11"
+    else:
+        raise RuntimeError(
+            "Can't compile with a layout larger than 3x3 with minimal caching on"
+        )
+
+
+def determine_rank_is_compiling(rank: int, partitioner: CubedSpherePartitioner) -> bool:
+    """Determines if a rank needs to be a compiling one
+
+    Args:
+        rank (int): current rank
+        partitioner (CubedSpherePartitioner): partitioner object
+
+    Returns:
+        bool: True if the rank is a compiling one
+    """
+    top_tile_equivalent = compiling_equivalent(rank, partitioner)
+    return rank == top_tile_equivalent
+
+
+def block_waiting_for_compilation(comm, compilation_config: CompilationConfig) -> None:
+    """block moving on until an ok is received from the compiling rank
+
+    Args:
+        comm (MPI.Comm): communicator over which the ok is sent
+        stencil_config (CompilationConfig): holding communicator and rank information
+    """
+    if comm and comm.Get_size() > 1:
+        compiling_rank = compilation_config.compiling_equivalent
+        _ = comm.recv(source=compiling_rank)
+
+
+def unblock_waiting_tiles(comm) -> None:
+    """sends a message to all the ranks waiting for compilation to finish
+
+    Args:
+        comm (MPI.Comm): communicator over which the ok is sent
+    """
+    rank = comm.Get_rank()
+    size = comm.Get_size()
+    if comm and size > 1:
+        for tile in range(1, 6):
```

Comment:
Is this 0 to 5 or 1 to 6?

Reply:
so you are on tile 0, so you don't want to send to yourself. So this is [1,6) so we are sending to all other tiles